tests/unit/s2n_connection_serialize_test.c

```diff
@@ -473,6 +490,271 @@ int main(int argc, char **argv)
+            EXPECT_SUCCESS(s2n_stuffer_rewrite(&cb_session_data));
+
+            /* Negotiate successful resumption connection */
+            EXPECT_SUCCESS(s2n_negotiate_test_server_and_client(server_conn, new_client_conn));
+            EXPECT_FALSE(IS_FULL_HANDSHAKE(new_client_conn));
+            EXPECT_FALSE(IS_FULL_HANDSHAKE(server_conn));
+        };
+
+        /* Server is serialized. Can write a session ticket after deserialization. */
+        {
+            DEFER_CLEANUP(struct s2n_connection *client_conn = s2n_connection_new(S2N_CLIENT),
+                    s2n_connection_ptr_free);
+            EXPECT_NOT_NULL(client_conn);
+            DEFER_CLEANUP(struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER),
+                    s2n_connection_ptr_free);
+            EXPECT_NOT_NULL(server_conn);
+
+            /* Wipe ticket from previous test */
+            EXPECT_SUCCESS(s2n_stuffer_wipe(&cb_session_data));
+
+            EXPECT_SUCCESS(s2n_connection_set_config(client_conn, resumption_config));
+            EXPECT_SUCCESS(s2n_connection_set_config(server_conn, resumption_config));
+
+            DEFER_CLEANUP(struct s2n_test_io_stuffer_pair io_pair = { 0 }, s2n_io_stuffer_pair_free);
+            EXPECT_OK(s2n_io_stuffer_pair_init(&io_pair));
+            EXPECT_OK(s2n_connections_set_io_stuffer_pair(client_conn, server_conn, &io_pair));
+
+            EXPECT_SUCCESS(s2n_negotiate_test_server_and_client(server_conn, client_conn));
+            EXPECT_EQUAL(s2n_connection_get_actual_protocol_version(server_conn), S2N_TLS13);
+
+            /* Get the client to do a read to pick up the session ticket. We won't use this one;
+             * we're interested in the tickets the server sends after the transfer. */
+            EXPECT_OK(s2n_send_and_recv_test(server_conn, client_conn));
+            EXPECT_SUCCESS(s2n_stuffer_wipe(&cb_session_data));
+
+            /* Serialize the server */
+            uint8_t buffer[S2N_SERIALIZED_CONN_TLS12_SIZE] = { 0 };
+            EXPECT_SUCCESS(s2n_connection_serialize(server_conn, buffer, sizeof(buffer)));
+
+            /* Initialize new server connection and deserialize session */
+            DEFER_CLEANUP(struct s2n_connection *new_server_conn = s2n_connection_new(S2N_SERVER),
+                    s2n_connection_ptr_free);
+            EXPECT_NOT_NULL(new_server_conn);
+            EXPECT_SUCCESS(s2n_connection_deserialize(new_server_conn, buffer, sizeof(buffer)));
+
+            /* Re-initialize IO pipes */
+            EXPECT_OK(s2n_connections_set_io_stuffer_pair(client_conn, new_server_conn, &io_pair));
+
+            /* The server automatically sends a ticket right after the handshake completes. However,
+             * we're interested in what happens after the server has been deserialized. Therefore we
+             * add another ticket for the server to send. */
+            EXPECT_SUCCESS(s2n_connection_add_new_tickets_to_send(new_server_conn, 1));
```

Comment:
Is this necessary? I would have expected the deserialized connection to automatically send its own session ticket, regardless of whether the original connection sent one.

Reply:
That is a good point. You're right, the number of tickets sent isn't serialized, so the deserialized server will send a new ticket when it gets the chance. I guess that makes sense, although, I definitely didn't expect that to be the case.